### ISSUE TITLE
Wrong Results for FP16 Models in CUDAExecutionProvider and TensorRTExecutionProvider

### ISSUE THREAD

**Issue by anogkongda** [microsoft/onnxruntime#12726]
**Describe the bug**
A clear and concise description of what the bug is. To avoid repetition please make sure this is not one of the known issues mentioned on the respective release page.

The model seems to "output zeros" for some MatMul operators for FP16 models sometimes, which leads to wrong results for output.
Many rows of the output matrix is equal to the "bias" of the fully-connected layer.

**Urgency**
As soon as possible please.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): CentOS7
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: can be reproduced on both  1.8.2 & 1.12.1
- Python version: 3.8.11
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source): 7.3.1 for ORT 1.8.2 and 8.3 for 1.12.1
- CUDA/cuDNN version: 11.0/8.0
- TensorRT version: 7.2.1 for ORT 1.8.2 and 8.4 for ORT 1.12.1
- GPU model and memory: NVIDIA T4 16G

**To Reproduce**
Run this python script several times, we can find the output of diffenent backends.

import torch
import torch.nn as nn
import onnxruntime as ort
import pdb
import numpy as np

xs = 100 * torch.randn(1,100,256).cuda()

model = nn.Sequential(
    nn.LayerNorm(256, eps=1e-12),
    nn.Linear(256,2048)
).cuda()

model.eval()

onnx_mdl = "linear.onnx"
onnx_mdl_fp16 = "linear_fp16.onnx"
dynamic_axes = {"x": [0,1]}
torch.onnx.export(model,
    xs,
    onnx_mdl,
    dynamic_axes=dynamic_axes,
    input_names=["x"],

)

model_16 = model.half()

xs_fp16 = xs.half()
y_torch = model_16(xs_fp16)
torch.onnx.export(model_16,
    xs_fp16,
    onnx_mdl_fp16,
    dynamic_axes=dynamic_axes,
    input_names=["x"],

)

tensorRT_providers_fp16 = [
    ('TensorrtExecutionProvider', {
        'device_id': 0,
        'trt_max_workspace_size': 2147483648,
        'trt_fp16_enable': True,
        'trt_dla_enable': False,
        'trt_engine_cache_enable': False,
        'trt_engine_cache_path':'./trtcache', 
    }),
    ('CUDAExecutionProvider', {
        'device_id': 0,
    })
]

session_opts = ort.SessionOptions()

session = ort.InferenceSession(onnx_mdl, session_opts, providers=tensorRT_providers_fp16)
session_cuda = ort.InferenceSession(onnx_mdl_fp16, session_opts, providers=['CUDAExecutionProvider'])
xs_npy = xs.detach().cpu().numpy()
ort_inputs = {
    "x": xs_npy
}
ort_inputs_fp16 = {
    "x": xs.detach().cpu().half().numpy()
}
y_ort = session.run(None, ort_inputs)
y_ort_cuda = session_cuda.run(None, ort_inputs_fp16)
print("Output of TensorRT FP16")
print(y_ort[0])
print("================")
print("Output of CUDA FP16")
print(y_ort_cuda[0])
print("================")
print("Output of pytorch FP16")
print(y_torch)
print("================")
print("The bias tensor of model")
print( model_16[1].state_dict()['bias'])
print(f"Sum of TRT_FP16: {np.sum(y_ort[0])}")
print(f"Sum of CUDA_FP16: {np.sum(y_ort_cuda[0])}")
print(f"Sum of TORCH_16: {torch.sum(y_torch)}")

**Expected behavior**

The output of CUDA Execution Provider and TensorRT should be close to Pytorch.
But now they outputs the bias.
**Screenshots**
The outputs are wrong for the same model and input with different backends

<img width="651" alt="image" src="https://user-images.githubusercontent.com/22041196/186662206-bb5ab791-17b9-4381-b499-0df1ed699657.png">

**Comment by RandySheriffH** [comment]
@chilo-ms for some insights

**Comment by jywu-msft** [comment]
as another data point, are you able to run the FP32 model with TensorRT EP (and set trt_fp16_enable to true)

**Comment by anogkongda** [comment]
> as another data point, are you able to run the FP32 model with TensorRT EP (and set trt_fp16_enable to true)
The python code mentioned above leverages the FP32 onnx model with trt_fp16_enable=true, the problem exists.

**Comment by anogkongda** [comment]
For pytorch -> ONNX,  the layernorm module would be splitted into several operaters such as reduce sum, div etc., which may easily bring about FP16 overflow.
We can implement a LayerNorm operator by Welfold Algorithm rather than standard (reduce sum -> divide).
or replace related operator by some safe operators in ONNX model.
